### PR TITLE
Use `std::chrono::steady_clock` in `getTickCount`

### LINKED ISSUE #6902: Add use of `std::chrono::steady_clock` into `cv::getTickCount` implementation when C++11 is available
- OpenCV => 3.x
- Operating System / Platform => any / c++11
- Compiler => any c++11
##### Detailed description

C++11 implements the standard ISO.20.11.7 in `std::chrono`, so it has much sense using this standard portable implementation inside `cv::getTickCount` when compiler supports it.

### PR DESCRIPTION
**Pull request by CSharperMantle**
This PR enables the use of `std::chrono::steady_clock` on supported compilers as a source to retrieve the current tick count and clock frequency in a cross-platform way.

Resolves #6902.

### Pull Request Readiness Checklist

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [x] There is a reference to the original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

**Comment by CSharperMantle** [comment]
I am particularly concerned about possible name clashes with macros.

In the patch, I store the result of `__cplusplus >= 201103L` in a macro named `HAVE_CXX11`, but it is a common name among libraries. Are there any better ways to detect C++11? Any suggestions will be appreciated.

File: modules/core/src/system.cpp [hunk 1]
```diff
@@ -840,7 +850,10 @@ bool useOptimized(void)
 
 int64 getTickCount(void)
 {
-#if defined _WIN32 || defined WINCE
+#if HAVE_CXX11
```
Comment (alalek) [inline modules/core/src/system.cpp:853]: Use `CV_CXX11`: https://github.com/opencv/opencv/blob/3.4.18/modules/core/include/opencv2/core/cvdef.h#L637

Comment (CSharperMantle) [reply modules/core/src/system.cpp]: Done.

**Comment by CSharperMantle** [comment]
One job failed due to a non-existent repository, `CSharperMantle/opencv_contrib`, but I suppose that it is not necessary for me to fork such repository. Are there any actions I should take?

**Comment by alalek** [comment]
To merge this patch we need to ensure that there is no unwanted performance regression against the existed code.
So we need microbenchmark results on major systems: Linux, Windows, MacOSX.

**Comment by CSharperMantle** [comment]
@alalek, unfortunately I don't an macOS X system, and my Linux system is quite atypical. May I ask whether the benchmarks should be carried out myself, or by your build farm?

**Comment by CSharperMantle** [comment]
## Benchmark 1: Windows

### Code

The following benchmark code utilizes macros to switch among different implementations of `getTickCount`.

```cpp
#include <chrono>
#include <cstdint>
#include <iostream>

#if defined(BENCH_IS_WINDOWS)
#include <windows.h>
#elif defined(BENCH_IS_MACH)    
#include <mach/mach_time.h>
#elif defined(BENCH_IS_UNIX)
#include <time.h>
#else
#include <sys/time.h>
#endif

#define RUN_COUNT 10000000

std::int64_t getTickCount()
{
#if defined(BENCH_IS_CXX11)
    std::chrono::steady_clock::time_point now = std::chrono::steady_clock::now();
    return (std::int64_t)now.time_since_epoch().count();
#elif defined(BENCH_IS_WINDOWS)
    LARGE_INTEGER counter;
    QueryPerformanceCounter(&counter);
    return (std::int64_t)counter.QuadPart;
#elif defined(BENCH_IS_MACH)
    return (std::int64_t)mach_absolute_time();
#elif defined(BENCH_IS_UNIX)
    struct timespec tp;
    clock_gettime(CLOCK_MONOTONIC, &tp);
    return (std::int64_t)tp.tv_sec * 1000000000 + tp.tv_nsec;
#else
    struct timeval tv;
    gettimeofday(&tv, NULL);
    return (std::int64_t)tv.tv_sec * 1000000 + tv.tv_usec;
#endif
}

int main(int, char **)
{
    volatile std::int64_t tmp = 0;

    std::cout << "Defined macros: ";
#if defined(BENCH_IS_CXX11)
    std::cout << "BENCH_IS_CXX11 ";
#elif defined(BENCH_IS_WINDOWS)
    std::cout << "BENCH_IS_WINDOWS ";
#elif defined(BENCH_IS_MACH)
    std::cout << "BENCH_IS_MACH ";
#elif defined(BENCH_IS_UNIX)
    std::cout << "BENCH_IS_UNIX ";
#else
    std::cout << "<none> ";
#endif
    std::cout << '\n';

    auto before = std::chrono::steady_clock::now();
    for (size_t i = 0; i < RUN_COUNT; i++)
    {
        tmp = getTickCount();
    }
    auto after = std::chrono::steady_clock::now();
    auto duration = std::chrono::duration_cast<std::chrono::nanoseconds>(after - before);

    std::cout << "Total time (ns): " << duration.count()
              << ", Avg time (ns): " << duration.count() / RUN_COUNT
              << '\n';

    return 0;
}

```

### Environment

* **OS:** Windows 10.0.10586
* **CPU:** Intel Core i3-4010U @ 1.70GHz, 1700 Mhz, 2C4T
* **Compiler:** g++.exe (x86_64-posix-seh-rev0, Built by MinGW-W64 project) 8.1.0

**Compiling commands:**

```sh
g++ -std=c++11 -DBENCH_IS_UNIX -O0 -o bench_unix.exe .\bench_getTickCount.cpp
g++ -std=c++11 -DBENCH_IS_CXX11 -O0 -o bench_cxx11.exe .\bench_getTickCount.cpp
g++ -std=c++11 -DBENCH_IS_WINDOWS -O0 -o bench_windows.exe .\bench_getTickCount.cpp
g++ -std=c++11 -O0 -o bench_default.exe .\bench_getTickCount.cpp
```

### Result

|  | BENCH_IS_CXX11 | BENCH_IS_UNIX | _BENCH_IS_WINDOWS_ | **&lt;none&gt;** |
|---|---|---|---|---|
| Total (ns) | 6618392421 | 6523856314 | _5999175594_ | **970594447** |
| Avg (ns) | 661 | 652 | _599_ | **97** |

The italicized column is the default implementation originally used on such platform; bold column is the fastest.

**Raw:**

```plain
PS F:\Workspace\bench> .\bench_cxx11.exe
Defined macros: BENCH_IS_CXX11
Total time (ns): 6618392421, Avg time (ns): 661
PS F:\Workspace\bench> .\bench_unix.exe
Defined macros: BENCH_IS_UNIX
Total time (ns): 6523856314, Avg time (ns): 652
PS F:\Workspace\bench> .\bench_windows.exe
Defined macros: BENCH_IS_WINDOWS
Total time (ns): 5999175594, Avg time (ns): 599
PS F:\Workspace\bench> .\bench_default.exe
Defined macros: <none>
Total time (ns): 970594447, Avg time (ns): 97
PS F:\Workspace\bench>
```

**Review by asmorkalov — APPROVED** [review]

**Pull request merged**